### ISSUE TITLE
schema: Allow FieldSerializer to be called on AnyOff members

### ISSUE THREAD

**Issue by smyrman** [rs/rest-layer#182]
### The problem
UPDATE: Let's split out schema.AllOf as a new issue, the solution is to different.

If a `FieldValidator` that implements the `FieldSerializer` interface is wrapped into a `schema.AnyOff` or `schema.AllOff`, the `FieldSerializer` implementation would be ignored, and the field would return it's internal db value instead of the intended serialized one. E.g.

```go
var s = schema.Schema{
    Fields: {
        "myField": {
            Validator: &schema.AnyOf{
                &schema.Null{},
                &CustomValidator{}, // Assumed to implement FieldSerialzer.
            },
        },
    },
}
```
The FieldSerializer implemented by `CustomValidator` would never be detected.

It's very obvious _why_ this does not work since there is no generic way in which the AnyOff or AllOff could determine which FieldSerializer to call (with a possible exception for the AllOff case).

### Possible solution
We might want some time to think of the best solution, but here is at least an initial suggestion.

One possible solution that should cover all cases and not do unnecessary work, would be to let the user _supply_ a serialize "picker" function  (or serialize function, but I will only demonstrate the first) to AnyOff / AllOff when relevant. E.g.:

```go
var s = schema.Schema{
    Fields: {
        "myField": {
            Validator: &schema.AnyOf{
                &schema.Null{},
                &CustomValidator{}, // Assumed to implement FieldSerialzer.
            }.WithSerializePicker(func(value interface{}) int {
                if value == nil {
                    return 0 
                }
                return 1
            }),
        },
    },
}
```

Where `WithSerializePicker` takes a function that returns the index of the FieldValidator that a value is assumed/defined to belong to, and returns a new FieldValidator type, e.g. something like `type WithSerializePicker struct{Validator []FieldValidator, SerializePicker SerializePickerFunc}` (can't be private type because the `jsonschema` package needs to type switch on it).

It is _not_ how ever a solution that _feels_ nice to use.

**Comment by smyrman** [comment]
Probably a nicer solution is to make AnyOf/AllOf a struct.

**Comment by rs** [comment]
For `AnyOff`, why can't we just use the first validator that match as a serializer?

**Comment by smyrman** [comment]
In the example case that wouldn't even necsarliy work, because `CustomValidator` may error on input it does not understand (nil in this case).

Here is another example where it defiantly won't work (some error catch code omitted):

```go
schema.AnyOf{
     &V1{},
     &V2{},
}

func (v V1) Validate(value interface{}) (interface{}, error) {
    return strconv.ParseFloat(value.(string), 64)
}

func (v V1) Serialize(value interface{}) (interface{}, error) {
    return fmt.Sprintf("%f", value)
}


func (v V2) Validate(value interface{}) (interface{}, error) {
    return strconv.ParseInt(value.(string), 10, 64)
}

func (v V1) Serialize(value interface{}) (interface{}, error) {
    return fmt.Sprintf("%d", value)
}
```

**Comment by smyrman** [comment]
We _could_ plausibly return the first Serialize result that doesn't error or fallback to the raw value.

It's not guaranteed to be the same FieldValidator that produced the value that ends-up Serialize the result, but that's not possible to guarantee anyway. Maybe that's good enough for the common use-cases, where anything else requires a custom FieldSerializer/FieldValodator to be implemented as a replacement for schema.AnyOf semantics.

**Comment by smyrman** [comment]
Let's just go with a variant of the latter for AnyOf. It's more or less symetric to the Validaror behaviour and it feels consistent (if documented).

It's easy enough for the user to do something more advanced when more control is neeeded over which serializer get's picked:
```
type CustomAnyOf struct{
    schema.AnyOf,
}


func (v CustomAnyOf) Serialize(value interface{}) (interface{}, error) {
    // switch on type or value to select serializer.
}
```

**Comment by smyrman** [comment]
Merged a solution in #183.
